sdk/core/Azure.Core/src/HttpMessage.cs

```diff
@@ -33,6 +33,7 @@ public HttpMessage(Request request, ResponseClassifier responseClassifier)
             Request = request;
             ResponseClassifier = responseClassifier;
             BufferResponse = true;
+            RetryContext = new RetryContext(DateTimeOffset.UtcNow);
```

Comment:
Since this is on the hot path, we should consider if there is a way to only allocate this on the first retry attempt while guaranteeing that it always gets done even in with a user provided custom policy.

Reply:
Another thing to consider - should we leverage the [internal typed propertybag ](https://github.com/Azure/azure-sdk-for-net/blob/75e00583fbaf2839d9bb6773a28db81f2fe4bb34/sdk/core/Azure.Core/src/HttpMessage.cs#L161-L177)rather than growing the size of the message object?

Reply:
Yes, we can do this. There is a slight issue of this not being thread-safe, but if we are making that trade-off for perf it should be fine.

Reply:
Based on feedback from @KrzysztofCwalina I've renamed this to ProcessingContext. This will be available as a property on HttpMessage even before retrying. I'm not sure the property bag approach will work here because we can't populate this lazily.